platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/TransactionHandlingHistory.java

```diff
@@ -0,0 +1,290 @@
+            return;
+        }
+
+        logger.info(STARTUP.getMarker(), "Log file found. Parsing previous history");
+
+        try (final BufferedReader reader = new BufferedReader(new FileReader(logFilePath.toFile()))) {
+            reader.lines().forEach(line -> addRoundToHistory(ConsistencyTestingToolRound.fromString(line)));
+        } catch (final IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Add a transaction to the list of seen transactions
+     *
+     * @param transaction the transaction to add
+     * @return an error message if the transaction has already been seen, otherwise null
+     */
+    @Nullable
+    private String addTransaction(final long transaction) {
+        if (!seenTransactions.add(transaction)) {
+            final String error =
+                    "Transaction with contents `" + transaction + "` has already been applied to the state";
+
+            logger.error(EXCEPTION.getMarker(), error);
+            return error;
+        }
+
+        return null;
+    }
+
+    /**
+     * Searches the history for a round which has the same round number as the new round
+     *
+     * @param newRound the round to search for a historical counterpart of
+     * @return the historical counterpart of the new round, or null if no such round exists
+     */
+    @Nullable
+    private ConsistencyTestingToolRound findHistoricalRound(final @NonNull ConsistencyTestingToolRound newRound) {
+        Objects.requireNonNull(newRound);
+
+        return roundHistory.stream()
+                .filter(oldRound -> newRound.compareTo(oldRound) == 0)
+                .findFirst()
+                .orElse(null);
+    }
+
+    /**
+     * Compare a newly received round with the historical counterpart. Logs an error if the new round isn't identical to
+     * the historical round
+     *
+     * @param newRound        the round that is being newly processed
+     * @param historicalRound the historical round that the new round is being compared to
+     * @return an error message if the new round doesn't match the historical round, otherwise null
+     */
+    @Nullable
+    private String compareWithHistoricalRound(
+            final @NonNull ConsistencyTestingToolRound newRound,
+            final @NonNull ConsistencyTestingToolRound historicalRound) {
+
+        Objects.requireNonNull(newRound);
+        Objects.requireNonNull(historicalRound);
+
+        if (!newRound.equals(historicalRound)) {
+            final String error =
+                    "Round " + newRound.roundNumber() + " with transactions " + newRound.transactionsContents()
+                            + " doesn't match historical counterpart with transactions "
+                            + historicalRound.transactionsContents();
+
+            logger.error(EXCEPTION.getMarker(), error);
+            return error;
+        }
+
+        return null;
+    }
```

Comment:
What if the node shuts down half way through handling a round, and some of that data makes its way to the disk? Will this code handle that case?

Reply:
As we discussed, this isn't considered. If a partial line is written to the logfile, and error will be logged at startup, but the partial line is otherwise ignored